File: packages/py-sdk/src/turbodocx_sdk/types/product.py
"""
TypedDict definitions for TurboQuote -- Product entity.
"""

from typing import Any, Dict, List, Optional, Union
from typing_extensions import TypedDict


class ProductImage(TypedDict, total=False):
    id: str
    productId: str
    fileId: str
    fileName: str
    fileType: str
    displayOrder: int
    imageData: str


class Product(TypedDict, total=False):
    id: str
    orgId: str
    name: str
    sku: Optional[str]
    description: Optional[str]
    detailedSpecification: Optional[str]
    internalNotes: Optional[str]
    categoryId: str
    listPrice: float
    cost: Optional[float]
    minimumOrderQuantity: int
    billingFrequency: str  # BillingFrequency literal
    currency: str  # Currency literal
    showInCatalog: bool
    isActive: bool
    createdBy: Optional[str]
    createdOn: str
    updatedOn: str
    images: List[Dict[str, Any]]  # ProductImage[]
    category: Dict[str, Any]


class CreateProductRequest(TypedDict, total=False):
    name: str
    listPrice: float
    billingFrequency: str  # BillingFrequency literal
    categoryId: str
    sku: str
    description: str
    detailedSpecification: str
    internalNotes: str
    cost: float
    minimumOrderQuantity: int
    currency: str  # Currency literal
    showInCatalog: bool
    images: List[Any]  # Array<string | bytes>


class UpdateProductRequest(TypedDict, total=False):
    name: str
    listPrice: float
    billingFrequency: str  # BillingFrequency literal
    sku: str
    description: str
    detailedSpecification: str
    internalNotes: str
    categoryId: str
    cost: float
    minimumOrderQuantity: int
    currency: str  # Currency literal
    showInCatalog: bool
    images: List[Any]  # Array<string | bytes>
    imageIdsToKeep: List[str]
    imageOrder: List[str]


class ListProductsOptions(TypedDict, total=False):
    limit: int
    offset: int
    query: str
    categoryIds: List[str]
    billingFrequency: str  # BillingFrequency literal
    currency: str  # Currency literal
    showInCatalog: bool


class ProductListResponse(TypedDict, total=False):
    results: List[Dict[str, Any]]
    totalRecords: int
    totalProducts: int
    activeProducts: int
    totalCategories: int
    catalogValue: float


# Dict mapping productId -> ProductImage | None
ProductPrimaryImagesResponse = Dict[str, Any]

File: packages/js-sdk/src/types/product.ts
/**
 * TypeScript types for TurboQuote — Product entity and request/response types
 */

import type { BillingFrequency, Currency, PaginationParams, PaginatedResponse } from './quote-shared';

// ============================================
// DOMAIN TYPES
// ============================================

export interface ProductImage {
  id: string;
  productId: string;
  fileId: string;
  fileName: string;
  fileType: string;
  displayOrder: number;
  imageData?: string;
}

export interface Product {
  id: string;
  orgId: string;
  name: string;
  sku: string | null;
  description: string | null;
  detailedSpecification: string | null;
  internalNotes: string | null;
  categoryId: string;
  listPrice: number;
  cost: number | null;
  minimumOrderQuantity: number;
  billingFrequency: BillingFrequency;
  currency: Currency;
  showInCatalog: boolean;
  isActive: boolean;
  createdBy: string | null;
  createdOn: string;
  updatedOn: string;
  images?: ProductImage[];
  category?: { id: string; name: string; categoryType: string };
}

// ============================================
// REQUEST TYPES
// ============================================

export interface CreateProductRequest {
  name: string;
  listPrice: number;
  billingFrequency: BillingFrequency;
  categoryId: string;
  sku?: string;
  description?: string;
  detailedSpecification?: string;
  internalNotes?: string;
  cost?: number;
  minimumOrderQuantity?: number;
  currency?: Currency;
  showInCatalog?: boolean;
  images?: Array<string | File | Buffer>;
}

export interface UpdateProductRequest {
  name?: string;
  listPrice?: number;
  billingFrequency?: BillingFrequency;
  sku?: string;
  description?: string;
  detailedSpecification?: string;
  internalNotes?: string;
  categoryId?: string;
  cost?: number;
  minimumOrderQuantity?: number;
  currency?: Currency;
  showInCatalog?: boolean;
  images?: Array<string | File | Buffer>;
  imageIdsToKeep?: string[];
  imageOrder?: string[];
}

export interface ListProductsOptions extends PaginationParams {
  categoryIds?: string | string[];
  billingFrequency?: BillingFrequency;
  currency?: Currency;
  showInCatalog?: boolean;
}

// ============================================
// RESPONSE TYPES
// ============================================

export interface ProductListResponse extends PaginatedResponse<Product> {
  totalProducts: number;
  activeProducts: number;
  totalCategories: number;
  catalogValue: number;
}

export interface ProductPrimaryImagesResponse {
  [productId: string]: ProductImage | null;
}
